src/ray/raylet/placement_group_resource_manager.cc

```diff
@@ -136,16 +148,103 @@ void OldPlacementGroupResourceManager::ReturnBundle(
+NewPlacementGroupResourceManager::NewPlacementGroupResourceManager(
+    std::shared_ptr<ClusterResourceScheduler> cluster_resource_scheduler_)
+    : cluster_resource_scheduler_(cluster_resource_scheduler_) {}
+
+bool NewPlacementGroupResourceManager::PrepareBundle(
+    const BundleSpecification &bundle_spec) {
+  auto iter = pg_bundles_.find(bundle_spec.BundleId());
+  if (iter != pg_bundles_.end()) {
+    if (iter->second->state_ == CommitState::COMMITTED) {
+      // If the bundle state is already committed, it means that prepare request is just
+      // stale.
+      RAY_LOG(INFO) << "Duplicate prepare bundle request, skip it directly.";
+      return true;
     } else {
-      iter++;
+      // If there was a bundle in prepare state, it already locked resources, we will
+      // return bundle resources.
+      ReturnBundle(bundle_spec);
+    }
+  }
+
+  std::shared_ptr<TaskResourceInstances> resource_instances =
+      std::make_shared<TaskResourceInstances>();
+  bool allocated = cluster_resource_scheduler_->AllocateLocalTaskResources(
+      bundle_spec.GetRequiredResources().GetResourceMap(), resource_instances);
+
+  if (!allocated) {
+    return false;
+  }
+
+  auto bundle_state =
+      std::make_shared<BundleTransactionState>(CommitState::PREPARED, resource_instances);
+  pg_bundles_[bundle_spec.BundleId()] = bundle_state;
+  bundle_spec_map_.emplace(bundle_spec.BundleId(), std::make_shared<BundleSpecification>(
+                                                       bundle_spec.GetMessage()));
+
+  return true;
+}
+
+void NewPlacementGroupResourceManager::CommitBundle(
+    const BundleSpecification &bundle_spec) {
+  auto it = pg_bundles_.find(bundle_spec.BundleId());
+  if (it == pg_bundles_.end()) {
+    // We should only ever receive a commit for a non-existent placement group when a
+    // placement group is created and removed in quick succession.
+    RAY_LOG(DEBUG)
+        << "Received a commit message for an unknown bundle. The bundle info is "
+        << bundle_spec.DebugString();
+    return;
+  } else {
+    // Ignore request If the bundle state is already committed.
+    if (it->second->state_ == CommitState::COMMITTED) {
+      RAY_LOG(INFO) << "Duplicate committ bundle request, skip it directly.";
+      return;
+    }
+  }
+
+  const auto &bundle_state = it->second;
+  bundle_state->state_ = CommitState::COMMITTED;
+
+  for (const auto &resource : bundle_spec.GetFormattedResources()) {
+    cluster_resource_scheduler_->AddLocalResource(resource.first, resource.second);
+  }
+}
+
+void NewPlacementGroupResourceManager::ReturnBundle(
+    const BundleSpecification &bundle_spec) {
+  auto it = pg_bundles_.find(bundle_spec.BundleId());
+  if (it == pg_bundles_.end()) {
+    RAY_LOG(INFO) << "Duplicate cancel request, skip it directly.";
+    return;
+  }
+  const auto &bundle_state = it->second;
+  if (bundle_state->state_ == CommitState::PREPARED) {
+    CommitBundle(bundle_spec);
```

Comment:
Can you write a comment why we commit here? Something like

```
Commit bundle first so that we can remove the bundle with consistent implementation.
```

Reply:
updated